lib/Lower/Mangler.cpp

```diff
@@ -62,15 +67,28 @@ Fortran::lower::mangle::mangleName(fir::NameUniquer &uniquer,
             return uniquer.doProgramEntry().str();
           },
           [&](const Fortran::semantics::SubprogramDetails &) {
-            auto &cb{symbol->name()};
-            auto modNames{moduleNames(symbol->scope())};
-            return uniquer.doProcedure(modNames, hostName(symbol->scope()),
-                                       toStringRef(cb));
+            const auto *scope{&symbol->owner()};
+            auto &symbolName{symbol->name()};
+            // Separate module subprograms must be mangled according to the
+            // scope where they were declared (the symbol we have is the
+            // definition).
+            if (symbol->attrs().test(Fortran::semantics::Attr::MODULE) &&
+                scope->IsSubmodule()) {
+              // FIXME symbol from MpSubprogramStmt do not seem to have
+              // Attr::MODULE set.
+              const auto *iface{scope->parent().FindSymbol(symbolName)};
+              assert(iface && "Separate module proc must be declared");
+              scope = &iface->owner();
+            }
+            auto modNames{moduleNames(*scope)};
+            return uniquer.doProcedure(modNames, hostName(*scope),
+                                       toStringRef(symbolName));
           },
           [&](const Fortran::semantics::ProcEntityDetails &) {
+            const auto &scope{symbol->owner()};
```

Comment:
I'm wondering if it might be a little easier to follow if `moduleName()` and `hostName()` took the symbol and then did the `symbol->owner()` deref internally. There is some quite subtle prep work required before calling those functions going on here.

Reply:
Good point, done, There is still something a bit ugly with the `Fortran::semantics::Attr::MODULE` stuffs that require a scope, but Peter just made a utility function that I will use once we have it.